applogic/src/api/markdown.rs

```diff
@@ -0,0 +1,690 @@
+    pub range: (u32, u32),
+    pub element: InlineElement,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+#[frb]
+pub struct RangedBlockElement {
+    pub range: (u32, u32),
+    pub element: BlockElement,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+#[frb]
+pub enum BlockElement {
+    Paragraph(Vec<RangedInlineElement>),
+    Heading(Vec<RangedInlineElement>),
+    Quote(Vec<RangedBlockElement>),
+    UnorderedList(Vec<Vec<RangedBlockElement>>), // Each item has multiple block elements
+    OrderedList(u64, Vec<Vec<RangedBlockElement>>), // Each item has multiple block elements
+    Table {
+        head: Vec<Vec<RangedBlockElement>>,
+        rows: Vec<Vec<Vec<RangedBlockElement>>>,
+    },
+    HorizontalRule,
+
+    /// If code blocks are indented, each line is a separate String
+    CodeBlock(Vec<((u32, u32), String)>), // Range and String
+
+    Error(String),
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+#[frb]
+pub enum InlineElement {
+    Text(String),
+    Code(String),
+    Link {
+        dest_url: String,
+        children: Vec<RangedInlineElement>,
+    },
+    Bold(Vec<RangedInlineElement>),
+    Italic(Vec<RangedInlineElement>),
+    Strikethrough(Vec<RangedInlineElement>),
+    Spoiler(Vec<RangedInlineElement>),
+    Image(String),
+    TaskListMarker(bool),
+    //UserMention(String),
+    //RoomMention(String),
+    //Video,
+    //Audio,
+    //Voice,
+    //Meeting,
+    //File,
+}
+
+impl MessageContent {
+    pub fn error(message: String) -> Self {
+        Self {
+            content: vec![RangedBlockElement {
+                range: (0, u32::try_from(message.chars().count()).unwrap()),
+                element: BlockElement::Error(message),
+            }],
+        }
+    }
+
+    #[frb(sync)]
+    pub fn try_parse_markdown_raw(string: Vec<u8>) -> Result<Self> {
+        Self::try_parse_markdown(&String::from_utf8(string).unwrap())
+    }
+
+    pub fn try_parse_markdown(string: &str) -> Result<Self> {
+        let parsed = Parser::new_ext(
+            string,
+            // Do not enable Options::ENABLE_GFM, it activates special blockquotes which are not part of the GFM spec https://github.com/orgs/community/discussions/16925
+            Options::ENABLE_STRIKETHROUGH | Options::ENABLE_TABLES | Options::ENABLE_TASKLISTS,
+        )
+        .into_offset_iter();
+        let mut result = Vec::new();
+        let mut iter = parsed
+            .map(|(element, range)| {
+                (
+                    element,
+                    u32::try_from(range.start).unwrap()..u32::try_from(range.end).unwrap(),
+                )
+            })
+            .peekable();
+
+        while iter.peek().is_some() {
+            result.push(parse_block_element(&mut iter, 1)?);
+        }
+
+        Ok(Self { content: result })
+    }
+}
+
+fn parse_block_element<'a, I>(iter: &mut Peekable<I>, depth: usize) -> Result<RangedBlockElement>
+where
+    I: Iterator<Item = (Event<'a>, Range<u32>)>,
```

Comment:
Is it possible to operate over a a struct with fields here instead of a tuple? Field names would make the code below a little more readable than using `.0` and `.1`. Same in the other parse functions below.